django/db/models/constraints.py

```diff
@@ -1,54 +1,105 @@
         sql = self.constraint_sql(model, schema_editor)
-        return schema_editor.sql_create_check % {
+        return schema_editor.sql_create_constraint % {
             'table': schema_editor.quote_name(model._meta.db_table),
-            'check': sql,
+            'name': schema_editor.quote_name(self.name),
+            'constraint': sql,
         }
 
     def remove_sql(self, model, schema_editor):
         quote_name = schema_editor.quote_name
-        return schema_editor.sql_delete_check % {
+        return schema_editor.sql_delete_constraint % {
             'table': quote_name(model._meta.db_table),
             'name': quote_name(self.name),
         }
 
+    def deconstruct(self):
+        path = '%s.%s' % (self.__class__.__module__, self.__class__.__name__)
+        path = path.replace('django.db.models.constraints', 'django.db.models')
+        return (path, (), {'name': self.name})
+
+    def clone(self):
+        _, args, kwargs = self.deconstruct()
+        return self.__class__(*args, **kwargs)
+
+
+class CheckConstraint(BaseConstraint):
+    def __init__(self, *, check, name):
+        self.check = check
+        super().__init__(name)
+
+    def constraint_sql(self, model, schema_editor):
+        query = Query(model)
+        where = query.build_where(self.check)
+        connection = schema_editor.connection
+        compiler = connection.ops.compiler('SQLCompiler')(query, connection, 'default')
+        sql, params = where.as_sql(compiler, connection)
+        params = tuple(schema_editor.quote_value(p) for p in params)
+        return schema_editor.sql_check_constraint % {'check': sql % params}
+
     def __repr__(self):
-        return "<%s: constraint='%s' name='%s'>" % (self.__class__.__name__, self.constraint, self.name)
+        return "<%s: check='%s' name=%r>" % (self.__class__.__name__, self.check, self.name)
 
     def __eq__(self, other):
         return (
             isinstance(other, CheckConstraint) and
             self.name == other.name and
-            self.constraint == other.constraint
+            self.check == other.check
         )
 
     def deconstruct(self):
-        path = '%s.%s' % (self.__class__.__module__, self.__class__.__name__)
-        path = path.replace('django.db.models.constraints', 'django.db.models')
-        return (path, (), {'constraint': self.constraint, 'name': self.name})
+        path, args, kwargs = super().deconstruct()
+        kwargs['check'] = self.check
+        return path, args, kwargs
 
-    def clone(self):
-        _, args, kwargs = self.deconstruct()
-        return self.__class__(*args, **kwargs)
+
+class UniqueConstraint(BaseConstraint):
+    def __init__(self, *, fields, name, defer=False):
+        if not fields:
+            raise ValueError('At least one field is required to define a unique constraint.')
+        self.fields = tuple(fields)
+        if defer:
+            self.defer = 'DEFERRABLE INITIALLY DEFERRED'
```

Comment:
There are three options for this [supported by postgres](https://www.postgresql.org/docs/current/static/sql-createtable.html):
* `NOT DEFERRABLE` - the default
* `DEFERRABLE INITIALLY IMMEDIATE`
* `DEFERRABLE INITIALLY DEFERRED`

I'm not sure what the best api for this is. [This comment](https://code.djangoproject.com/ticket/20581#comment:4) suggests two flags, `deferrable` and `initially_deferred`, but this feels clunky to me, since the second flag is meaningless if the first is `False`. I'm currently leaning towards an enum approach like Django's `on_delete` behaviour for `ForeignKey`:

* `constraints.NOT_DEFERRABLE`
* `constraints.DEFERRABLE_INITIALLY_IMMEDIATE`
* `constraints.DEFERRABLE_INITIALLY_DEFERRED`

I'm also wondering about implementing support for [`SET CONSTRAINTS`](https://www.postgresql.org/docs/current/static/sql-set-constraints.html) as part of the `transaction.atomic` context manager to increase the usefulness of `INITIALLY IMMEDIATE` constraints.